src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs

```diff
@@ -159,9 +183,10 @@ public bool Invoke (MethodProxy calledMethod, MultiValue instanceValue, IReadOnl
 				return true;
 			}
 
-			// For some intrinsics we just use the return value from the annotations.
-			if (returnValue == null)
-				returnValue = calledMethod.ReturnsVoid () ? MultiValueLattice.Top : GetMethodReturnValue (calledMethod, returnValueDynamicallyAccessedMemberTypes);
+			if (returnValue == null && !calledMethod.ReturnsVoid ())
+				throw new InvalidOperationException ("No return value set for intrinsic");
```

Comment:
Given how many times we've been struggling with this (and thus the implied lack of test coverage) - I'm a bit nervous making it throw here (if linker throws, there's basically no workaround for trimming the app). Maybe we should assert and fallback to returning the annotated value - that should in theory lead to warnings which people report to us as noise (and we still catch the most glaring holes in our tests if we hit them).

On a related note (as this has been a recurring topic recently) - maybe we should use Debug builds of linker in our repos (runtime, MAUI, ...) when they run their Debug CI legs - that way we get better test coverage of the linker without the need to break app trimming with exceptions (we could basically use asserts more frequently without losing that much data).

Reply:
I left this as before, because some of the new intrinsics rely on the fall-back. Even though they don't return types, they currently get are tracked as ValueWithDynamicallyAccessedMembers, and then cause warnings in some cases, like MakeGenericMethod. I think these should be changed to track UnknownValue.